### PR TITLE
fix(autoresearch): bound RP PIO lane sizes by what the capture can actually hold (#4371)

### LINKED ISSUE #4371: runSingleTest advertises a lane cap it cannot honour: max_leds_per_lane uses a fixed /32 divisor for a geometry-dependent limit
`runSingleTest` bounds `laneSizes` with a constant that only holds for one chipset geometry. Above the real limit the guard passes the request through and capture fails **silently** — `passed: false`, `bytes: 0`, no error — which is the outcome the guard exists to prevent.

**`examples/AutoResearch/AutoResearchRemoteRunSingleTest.cpp:512`**

```cpp
const int max_leds_per_lane = mState->rx_buffer.size() / 32;
```

`rx_buffer` is `fl::span<uint8_t>` of 3300 bytes, so the advertised cap is `3300 / 32 = 103`.

## Measured on an RP2350W (PIO0, GPIO0 -> GPIO1 bridge)

Every row reproduced 3/3:

| geometry | advertised cap | actual usable | above it |
|---|---|---|---|
| `WS2812B-V5` (800 kHz) | 103 | 100–102 | silent failure |
| `WS2811-400KHZ` (400 kHz) | 103 | 60–64 | silent failure |

```
WS2811-400KHZ leds= 60 passed=True  bytes=720
WS2811-400KHZ leds= 65 passed=False bytes=0     <- accepted by the guard, fails in capture
WS2812B-V5    leds=100 passed=True  bytes=1200
WS2812B-V5    leds=103 passed=False bytes=0     <- the advertised maximum itself
WS2812B-V5    leds=110 -> {"error": "LaneSizeTooLarge", "message": "... exceeds max 103"}
```

## Two defects

**1. The advertised maximum does not work, even at 800 kHz.** 103 is accepted and then fails, 3/3. A caller who trusts the number the device reports gets silence rather than `LaneSizeTooLarge` — strictly the worse of the two failures, since the guard's whole purpose is to turn this into a clean error.

**2. A fixed divisor cannot express this limit.** Decoded LED data is 3 bytes per LED; 32 bytes per LED means the buffer holds raw capture samples, whose cost scales with how long the frame *lasts*. A 400 kHz frame is twice as long as an 800 kHz frame of the same LED count, so no single constant describes both. At 400 kHz the cap is ~70% too high, and everything from 65 to 103 LEDs is accepted and silently fails.

## Why this matters beyond the constant

The silent band is wide enough to be mistaken for a different bug. The AutoResearch default is 100 LEDs, which sits deep inside the 400 kHz failure band — so `--chipset ws2811-400` appears to fail on every driver, and I spent a while concluding the RX decoder could not handle 400 kHz at all. It handles it fine below the real ceiling. Anyone testing a slower chipset will hit the same false trail.

## Direction

`runSingleTest` already receives the timing (`"timing": "WS2811-400KHZ"`) and resolves it to `t1_ns/t2_ns/t3_ns` a few lines later for `make4PhaseTiming`, so the information for a geometry-aware bound is available at the same call site — the cap could scale with the bit period rather than being constant.

**I am deliberately not proposing a number.** Deriving the true bytes-per-LED needs the PIO_RX sample rate and whether it is fixed or tracks the bit period, which I could not establish black-box: no simple model fits both thresholds. Frame duration does not (3.09 ms fails while 3.60 ms passes), bit count does not (2400 bits pass while 1560 fail), and even inverse-proportionality does not — halving the bit rate moves the usable cap by 1.6x, not 2x. Whoever owns that capture path can size it properly; the table above pins what any correct answer has to reproduce.

The guard's shape is right — a bounds check returning a clear error. Only the bound is wrong, and changing the constant will not fix it.

Found while working #3899. Related: #3899 (RP2350 HIL matrix).

### PR DESCRIPTION
**Pull request by zackees**
`runSingleTest` bounded `laneSizes` with `mState->rx_buffer.size() / 32`. That measures the wrong thing twice over: `rx_buffer` is the decode **output** buffer, and 32 has no relation to the capture. It advertised 103 LEDs — and 101, 102 and 103 were all *accepted* and then failed in capture with `bytes: 0` and no error, which is the outcome a bounds check exists to prevent (#4371).

## Two ceilings, each binding in its own regime

This is why no single constant — and no single model — fit both chipsets.

**1. Edge pool.** The capture appends one `EdgeTime` per signal phase into a fixed pool: two phases per bit, eight bits per wire byte. So `kRpPioRxEdgeCapacity / 16` = **300 wire bytes**.

**2. Sample-time budget.** The sampler packs 32 samples per DMA word at 20 MHz, and `mDmaWordCount = (mCapacity + 1) / 2 + kPioRxDmaTailWords` is pinned at **2465** for every frame large enough for either ceiling to matter. That is a constant quantity of *wall clock*, not of data:

```
2465 words x 32 samples x 50 ns = 3.944 ms
  - 100 us trailing idle (RxChannelConfig::signal_range_max_ns)
  -  64 us arm-to-first-edge gap
  = 3.78 ms for the frame
```

At a 1225 ns bit period the pool binds first (300 bytes, 2.9 ms). At 2500 ns the clock binds first (3.78 ms, 189 bytes). Each ceiling is invisible in the other's regime, which is why 800 kHz parts all reported 100 and the one 400 kHz part reported 63.

The two were measured apart: **192 wire bytes / 3072 phases passes 3/3 at 800 kHz and fails 3/3 at 400 kHz.** Same size, same phase count, opposite verdicts — so constraint 2 is about time and constraint 1 is about size.

## And it has to be encoder-aware

A UCS7604 lane is more than twice a WS2812 lane of the same length — 15-byte preamble, 16-bit channels, padding to a multiple of three — so it tops out at **47** LEDs where WS2812 tops out at 100. A guard phrased in LEDs cannot be right for both; one phrased in wire bytes is right for both.

## The change

- `fl::validation::rpPioMaxWireBytes()` returns the smaller of the two ceilings, from the sampler's own constants.
- The check moves to just after the timing and encoder are resolved, because it needs both. The old guard stays as a coarse pre-allocation backstop.
- `kPioRxClockHz`, `kPioRxSamplesPerDmaWord` and `kPioRxDmaTailWords` move into `rx_pio_channel.h` so the bound reads the real constants rather than copies; `collectDurations()` now uses the named samples-per-word constant instead of a bare `32`.
- `ucs7604FrameBytes()` joins the encoder header for the same reason — `encodeUCS7604()` derives its padding from it, so the length callers predict and the length written cannot drift.

## Measured

RP2350W `2DCB876B587EA334`, PIO0, GPIO0 → GPIO1. Every boundary is now the hardware's:

| timing | last accepted | first rejected |
|---|---|---|
| WS2812B-V5 | 100, 4/4 | 101 — 303 B > 300 |
| WS2818 | 100, 4/4 | 101 — 303 B > 300 |
| WS2814 | 100, 4/4 | 101 — 303 B > 300 |
| WS2811-400KHZ | 63, 4/4 | 64 — 192 B > 189 |
| UCS7604-800KHZ | 47 | 48 — 303 B > 300 |

The silent band is empty for all five. The rejection now says what happened:

```
LaneSizeTooLarge: laneSizes[0] = 64 puts 192 bytes on the wire at
WS2811-400KHZ; PIO RX capture holds 189 (63 LEDs)
```

Host tests assert the same five ceilings against the production function rather than re-deriving them.

> UCS7604 still compares as *failed* on this branch — that is #4373, fixed separately in #4377; this branch is off master. It captures 141 bytes at 47 LEDs, `47 x 3` raw pixel bytes, exactly that defect's signature. The guard behaviour under test here is unaffected.

Fixes #4371

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01KkufoNxfnNRU9psT3R9F51

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved RP2040 and RP2350 validation for LED configurations exceeding available capture capacity.
  * Correctly accounts for RGBW data when calculating wire usage and maximum LED counts.
  * Improved UCS7604 frame sizing for RGB and RGBW configurations, including protocol padding.
  * Added support for wider-source RGB output with improved color handling.

* **Tests**
  * Added coverage for RP PIO capture limits across 800 kHz and 400 kHz LED protocols.
  * Added validation for boundary conditions and insufficient capture budgets.

### CODE REVIEW

**Comment by zackees** [comment]
Two of the four applied, two declined with reasons.

## Applied — gate on the RX backend, not the TX driver

Correct, and it was a live hole rather than a theoretical one. The ceiling belongs to whatever *captures*; on RP, `PLATFORM_DEFAULT` resolves to PIO RX whatever the TX driver is. Keyed off `driver_name`, a `UART0` request — captured by the same pool, subject to the same ceiling — skipped the check entirely.

Confirmed on hardware before the fix: `UART0` at 64 and 100 LEDs went straight through into exactly the silent capture failure this guard exists to prevent. After moving the check to follow the RX channel and testing `rx_channel_to_use->backend()`:

```
PIO0  WS2812B-V5      100 accepted 4/4   101 rejected (303 B > 300)
PIO0  WS2811-400KHZ    63 accepted 4/4    64 rejected (192 B > 189)
PIO0  UCS7604-800KHZ   47 accepted        48 rejected (303 B > 300)
UART0 WS2811-400KHZ    63 accepted        64 rejected (192 B > 189)
UART0 WS2811-400KHZ                      100 rejected (300 B > 189)
```

## Applied — one source for the sampler constants

Also correct: the tests restated `4801 / 64 / 32 / 50` and could have passed while production computed a different limit. They were copied because `rx_pio_channel.h` sits behind the RP build guard and the test runs on the host.

Fixed at the cause rather than the symptom — the four constants move to `fl/channels/rx/pio_geometry.h`, portable and outside that guard, so the device, the bound in `validation.h` and the host test all read the same definitions. Renamed with a `kRpPioRx` prefix on the way out, since they are no longer file-local.

## Declined — moving the new definitions out of `.h`

Both instances follow the established convention of the files they are in, and moving only the new ones would leave each header internally inconsistent.

`src/fl/chipsets/encoders/` defines `inline` non-template functions in `.h` as its normal pattern:

| file | inline non-template definitions |
|---|---|
| `encoder_utils.h` | 6 (`mapBrightness8to5`, `p9813FlagByte`, `lpd8806Encode`, `hd108GammaCorrect`, `hd108BrightnessHeader`, `lpd6803EncodeRGB`) |
| `pixel_iterator_adapters.h` | 5 |
| `pixel_iterator.h` | 5 |
| `ws2816.h` | 1 (`packWS2816Pixel`) |

`ucs7604FrameBytes` and `ucs7604BytesPerLed` are two arithmetic expressions with no dependencies, sitting directly above the template that consumes them — the same shape as all of the above.

`src/fl/channels/validation.h` is the same on master, before this PR: `useRmtInternalLoopback`, `resolveCaptureBackend` and `captureEdgeCapacity` are all `inline` definitions in that header, and `rpPioMaxWireBytes` is a peer of `captureEdgeCapacity` in both role and size.

I am not arguing the rule is wrong — if these headers should be split, that is a worthwhile cleanup, but it is a change to four existing functions in `validation.h` and a dozen across the encoder headers, not something to do halfway inside a fix for a capture bound. Happy to do it as its own PR if that is wanted.

`bash lint` clean, host suite 304/304 + 84/84.

**Comment by zackees** [comment]
Applied — correct, and reachable on this board.

The WS2812 branch of `rpPioWireFrameBytes()` returned `leds * 3` unconditionally. With `legacyRgbw`, or a chipset with automatic RGBW, `LegacyClocklessProxy` emits four bytes per LED — so a 100-LED RGBW frame is 400 wire bytes and the guard waved it through against a 300-byte ceiling. That is the silent failure this check exists to remove, surviving in the case I did not cover.

Both helpers now take the effective width, computed with the same predicate the comparison uses (`legacyLaneUsesRgbw`): the request's `legacyRgbw` flag, or `legacyClocklessChipsetHasAutomaticRgbw()` for the lane's chipset.

Measured on RP2350W `2DCB876B587EA334`, PIO0, GPIO0 → GPIO1:

| lane | last accepted | first rejected |
|---|---|---|
| legacy WS2812B (RGB) | 100, 4/4 | 101 — 303 B |
| legacy WS2814 (automatic RGBW) | **75**, 8/8 | **76 — 304 B** |
| legacy WS2812B + `legacyRgbw` | **75**, 4/4 | **76 — 304 B** |

75 × 4 = 300 bytes = 4800 phases, one under the 4801-slot pool; 76 × 4 = 304 bytes = 4864, over it. Predicted and measured agree to the LED on all three.

`bash lint` clean, host suite green.

**Comment by zackees** [comment]
@coderabbitai review

The findings on this PR were addressed in commits pushed after the last review, and the review that still carries `CHANGES_REQUESTED` predates them. Requesting a fresh pass so the state reflects the current branch — replies detailing each change are above.

File: examples/AutoResearch/AutoResearchRemoteRunSingleTest.cpp [hunk 1]
```diff
@@ -846,6 +917,58 @@ fl::json AutoResearchRemoteControl::runSingleTestImpl(const fl::json& args) {
...
+    // The `laneSizes` guard above bounds the request with
+    // `rx_buffer.size() / 32`, which measures the wrong thing: `rx_buffer` is
+    // the decode *output*, and 32 has no relation to the capture. It advertised
+    // 103 LEDs, and 101, 102 and 103 were all accepted and then failed in
+    // capture with no error at all -- the outcome a bounds check exists to
+    // prevent (FastLED#4371).
+    //
+    // The real ceiling needs the resolved timing and encoder, which is why the
+    // check lands here rather than with the rest of the argument validation.
+    if (driver_name == "PIO0" || driver_name == "PIO1" || driver_name == "PIO2") {
```
[the remark this answers is not in the document]
Comment (zackees) [reply examples/AutoResearch/AutoResearchRemoteRunSingleTest.cpp]: Correct and important, fixed in 2e98824. The bound was keyed off `driver_name` -- the transmitter -- when the ceiling belongs to whatever captures. On RP, `PLATFORM_DEFAULT` resolves to PIO RX whatever the TX driver is, so a UART0 request captured by the same pool and subject to the same ceiling skipped the check entirely, and a PIO TX pointed at another backend would have received a PIO-only rejection.

Confirmed on hardware before the fix: UART0 at 64 and 100 LEDs went straight through into the silent capture failure the guard exists to prevent. It now resolves the RX channel first and tests `rx_channel_to_use->backend()` against both `PLATFORM_DEFAULT` and `PIO`, and both paths are rejected with the message PIO0 already got.

File: examples/AutoResearch/AutoResearchRemoteRunSingleTest.cpp [hunk 2]
```diff
@@ -942,6 +1013,68 @@ fl::json AutoResearchRemoteControl::runSingleTestImpl(const fl::json& args) {
...
+            fl::kRpPioRxEdgeCapacity, fl::kRpPioRxDmaTailWords,
+            fl::kRpPioRxSamplesPerDmaWord,
+            1000000000u / fl::kRpPioRxClockHz,
+            fl::RxChannelConfig(pin_rx).signal_range_max_ns,
+            kArmingLeadInNs, bit_period_ns);
+
+        for (fl::size i = 0; i < lane_sizes.size(); i++) {
+            const fl::size leds = static_cast<fl::size>(lane_sizes[i]);
+            const fl::size wire_bytes =
+                rpPioWireFrameBytes(leds, timing_config.encoder);
```
[the remark this answers is not in the document]
Comment (zackees) [reply examples/AutoResearch/AutoResearchRemoteRunSingleTest.cpp]: Real hole, fixed in be08f51. The WS2812 branch of `rpPioWireFrameBytes()` returned `leds * 3u` unconditionally, so with `legacy_rgbw` (or a chipset with automatic RGBW) `LegacyClocklessProxy` emits four bytes per LED and a 100-LED frame is 400 wire bytes -- accepted against a 300-byte ceiling, which is exactly the silent failure this guard exists to remove. The effective RGBW width is now computed once per lane and passed to both wire-byte helpers, including the UCS7604 branch via `ucs7604FrameBytes`/`ucs7604BytesPerLed`.

File: tests/fl/channels/validation.cpp [hunk 3]
```diff
@@ -129,6 +129,72 @@ FL_TEST_CASE("ISR validation capture uses a frame-sized power-of-two buffer") {
...
+// The RP PIO sampler's own constants, so the expectations below are the
+// numbers the device actually runs with (rx_pio_channel.h). They are repeated
+// rather than included because that header is RP-only and this test builds on
+// the host.
+constexpr size_t kEdgeCapacity = 100u * 3u * 16u + 1u;  // 4801
+constexpr size_t kDmaTailWords = 64u;
+constexpr u32 kSamplesPerWord = 32u;
+constexpr u32 kSamplePeriodNs = 50u;   // 20 MHz
+constexpr u32 kIdleTailNs = 100000u;   // RxChannelConfig default
+constexpr u32 kArmingLeadInNs = 64000u;
```
[the remark this answers is not in the document]
Comment (zackees) [reply tests/fl/channels/validation.cpp]: Agreed, fixed in 2e98824. The sampler constants now live in `src/fl/channels/rx/pio_geometry.h` -- portable, outside the RP build guard -- and the device (`rx_pio_channel.cpp.hpp`, which now sets its clkdiv from `kRpPioRxClockHz`), the bound, and this host test all read the same four values. The duplicated literals are gone from the test.

File: src/fl/chipsets/encoders/ucs7604.h [hunk 4]
```diff
@@ -194,6 +194,28 @@ void encodeUCS7604_16bit_RGBW(InputIterator first, InputIterator last, OutputIte
     }
 }
 
+/// @brief Wire bytes each LED contributes to a UCS7604 frame
+inline size_t ucs7604BytesPerLed(UCS7604Mode mode, bool is_rgbw) FL_NO_EXCEPT {
+    if (mode == UCS7604Mode::UCS7604_MODE_8BIT_800KHZ) {
+        return is_rgbw ? 4u : 3u;
+    }
+    return is_rgbw ? 8u : 6u;
+}
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/chipsets/encoders/ucs7604.h:245]: Declining. `src/fl/chipsets/encoders/` is header-only -- there is no `.cpp.hpp` or `.impl.hpp` anywhere in that directory -- so "move the definitions to an implementation file" means inventing the first compiled implementation file the encoder tree has ever had, as a rider on a capture-bound fix.

Nor are these the first non-template definitions of their shape there. On master `encoder_utils.h` has six, `pixel_iterator.h` and `pixel_iterator_adapters.h` five each, `encoder_constants.h` three, `ws2816.h` one. `ucs7604BytesPerLed` is a two-branch size lookup and `ucs7604FrameBytes` is `15 + n * bytesPerLed`, rounded up to a multiple of 3.

There is also a cost to moving them. `encodeUCS7604` is a template in this same header and calls `ucs7604FrameBytes` to compute its own padding -- that is deliberate, so the size helper and the encoder cannot disagree about frame length. Behind a link-time call it stops inlining into every instantiation, on the chipset encoders, which are the paths this repo measures flash on.

File: src/fl/chipsets/encoders/ucs7604.h [hunk 5]
```diff
@@ -194,6 +194,28 @@ void encodeUCS7604_16bit_RGBW(InputIterator first, InputIterator last, OutputIte
...
+/// that must size a capture or predict whether a frame fits need this without
+/// encoding it first (FastLED#4371); `encodeUCS7604` computes its own padding
+/// from here so the two cannot drift.
+inline size_t ucs7604FrameBytes(size_t num_leds, UCS7604Mode mode,
+                                bool is_rgbw) FL_NO_EXCEPT {
+    constexpr size_t kPreambleLen = 15;
+    const size_t total = kPreambleLen + num_leds * ucs7604BytesPerLed(mode, is_rgbw);
+    // The protocol requires the total to be divisible by 3.
+    return total + (3u - (total % 3u)) % 3u;
+}
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/chipsets/encoders/ucs7604.h:259]: Same answer as the sibling thread for the two UCS7604 helpers -- `src/fl/chipsets/encoders/` has no implementation file to move them to, and five headers there already carry definitions of that shape.

On `src/fl/channels/validation.h#L132-165`: `rpPioMaxWireBytes` is the fourth `inline` free function in the `fl::validation` namespace in that header, after `useRmtInternalLoopback`, `resolveCaptureBackend` and `captureEdgeCapacity` -- all pre-existing on master, all pure functions over scalars, all defined in-class-style right there. It is the established shape of the file, and it is why the host test in `tests/fl/channels/validation.cpp` can call the production function directly instead of mirroring the arithmetic, which `agents/docs/cpp-standards.md` asks for explicitly.

Moving the fourth one alone would leave the namespace split across two files by nothing but merge date. If the `src/**/*.h` instruction is meant to bind here, the whole namespace should move in a dedicated PR.

**Comment by zackees** [comment]
Rebased onto master (now at the #4383/#4381/#4367/#4318 merges). Two conflicts, both additive:

- `src/fl/chipsets/encoders/ucs7604.h` -- master grew the `wide_source` / `FL_PLATFORM_HAS_TINY_MEMORY` block in `encodeUCS7604` (#4387) at the same place this PR replaced the inline bytes-per-LED arithmetic with a call to `ucs7604FrameBytes()`. Kept both: the TINY guard is untouched, and the padding still derives from the shared helper so the length callers predict and the length written cannot drift.
- `examples/AutoResearch/AutoResearchRemoteRunSingleTest.cpp` -- master added `StandingRxChannelSwap` (#4374) where this PR added the RP PIO wire-byte helpers. Unrelated additions at the same offset; both kept.

Verified locally on the rebased head: `bash lint` clean, `bash test --cpp` 305/305 unit and 84/84 examples.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9
